### PR TITLE
Grammar: Paths

### PR DESCRIPTION
**Pull request by ehuss**
This includes a variety of things all related to paths:

- Paths
- Path expressions
- Method call expression
- Attribute clarifications
- Visibility fixes

### CODE REVIEW

File: src/paths.md [hunk 1]
```diff
@@ -1,72 +1,137 @@
...
+S::f();  // Calls the inherent impl.
+<S as T1>::f();  // Calls the T1 trait function.
+<S as T2>::f();  // Calls the T2 trait function.
+```
+
+### Paths in types
+
+> **<sup>Syntax</sup>**\
+> _TypePath_ :\
+> &nbsp;&nbsp; `::`<sup>?</sup> _TypePathSegment_ (`::` _TypePathSegment_)<sup>\*</sup>
```
Comment (matthewjasper) [inline src/paths.md]: Type paths also include `Fn(usize) -> isize` (and `std::ops::Fn::(usize) -> isize`)

File: src/paths.md [hunk 2]
```diff
@@ -1,72 +1,137 @@
...
-mod a {
-    pub fn foo() {}
+(0..10).collect::<Vec<_>>();
+Vec::<u8>::with_capacity(1024);
+```
+
+## Qualified paths
+
+> **<sup>Syntax</sup>**\
+> _QualifiedPathInExpression_ :\
```
Comment (matthewjasper) [inline src/paths.md]: Should there also be a  _QualifiedPathInType_?

File: src/paths.md [hunk 3]
```diff
@@ -1,72 +1,137 @@
...
+```
+
+### Paths in types
+
+> **<sup>Syntax</sup>**\
+> _TypePath_ :\
+> &nbsp;&nbsp; `::`<sup>?</sup> _TypePathSegment_ (`::` _TypePathSegment_)<sup>\*</sup>
+>
+> _TypePathSegment_ :\
+> &nbsp;&nbsp; _PathIdentSegment_ (`::`<sup>?</sup> [_Generics_])<sup>?</sup>
```
Comment (matthewjasper) [inline src/paths.md]: There's also a variant that allows specifying associated items (i.e. `Iterator<Item=i32>`). I'm not sure if having two variants here is better or worse that specifying when it's allow in the prose.

File: src/paths.md (hunk 1 above)
Comment (ehuss) [reply src/paths.md]: Oops!  I got distracted researching `unboxed_closures` feature and forgot to finish this!

File: src/paths.md (hunk 3 above)
Comment (ehuss) [reply src/paths.md]: I'm not sure if I'm following.  Is that form not part of the generics (_TypeParam_ in [generics](https://github.com/rust-lang-nursery/reference/blob/master/src/items/generics.md))?

Comment (matthewjasper) [reply src/paths.md]: Yes, I must have missed that

Comment (ehuss) [reply src/paths.md]: I realized I had the wrong rule listed, updated.

**Comment by ehuss** [comment]
I didn't document any of the 2018 behavior, but I guess the current nightly should in theory be the first stable release of 2018, so I'll add that soon.  For some reason I was thinking there was another release in-between, time flies!

AFAIK, the reference currently doesn't have *any* 2018 documentation.  I'll need to spend some time to try to figure out a clear way to show the differences.  #281 mentioned using boxes, but I'm not sure if that really fits (should the box document the new behavior or the old?).  I'll experiment a little.

**Comment by Havvy** [comment]
There are four blockquotes that say "Edition Differences". They're mainly like that because no thought has been put into it. If you can think of a better way, feel free to change them to that.

**Comment by ehuss** [comment]
> There are four blockquotes that say "Edition Differences".

Ah, my branches were out of date so I didn't see that.  

I've updated with some information about 2018 paths.  I explicitly skipped over the `extern crate` changes, as I'd like that in a separate PR.  I waffled back and forth on the details and wording, I have no idea if it's clear at all.

**Pull request merged**